thorlcr/slave/traceslave.hpp

```diff
@@ -0,0 +1,157 @@
+
+#ifndef TRACESLAVE_HPP
+#define TRACESLAVE_HPP
+
+#include "slave.hpp"
+#include "jqueue.tpp"
+#include "jthread.hpp"
+
+
+class CRowRingBuffer:  public SafeRingBufferOf<const void>
+{
+    virtual void releaseBuffer(const void *e)
+    {
+        DBGLOG("releaseBuffer: %p, ordinality %u",e, ordinality());
+        ReleaseThorRow(e);
+    };
+public:
+    virtual ~CRowRingBuffer() {};
+};
+
+class CRowTracerThread : extends Thread
+{
+    CRowRingBuffer *rowRingBuffer;
+    IHThorArg *helper;
+    Semaphore waitSemaphore;
+    bool continueTracing;
+    unsigned intervalInMSecs;
+    int activityId;
+public:
+    CRowTracerThread(IHThorArg *_helper, unsigned sizeRingBuffer, unsigned _intervalInMSecs, unsigned _activityId)
+    : Thread("RowTracerThread"), helper(_helper),
+      intervalInMSecs(_intervalInMSecs), activityId(_activityId)
+    {
+        rowRingBuffer = new CRowRingBuffer;
+        rowRingBuffer->ensure(sizeRingBuffer);
+    };
+
+    virtual ~CRowTracerThread()
+    {
+        DBGLOG ("TRACE: Destructor ActivityId: %d)", activityId);
+        delete rowRingBuffer;
+    }
+    virtual int run()
+    {
+        continueTracing = true;
+        while (continueTracing)
+        {
+            const void *row;
+            waitSemaphore.wait(intervalInMSecs);
+            DBGLOG ("TRACE: RUN (ActivityId: %d, %u)", activityId, rowRingBuffer->ordinality());
+
+            for (int maxRows=10; maxRows; --maxRows)
+            {
+                row = rowRingBuffer->dequeue();
+                if (!row) break;
+
+                CommonXmlWriter xmlwrite(XWFnoindent);
+                helper->queryOutputMeta()->toXML((const byte *) row, xmlwrite);
+                DBGLOG ("TRACE: (ActivityId: %d, %p, %u) <ROW>%s</ROW>", activityId, row, rowRingBuffer->ordinality(), xmlwrite.str());
+                ReleaseThorRow(row);
+            }
+        };
+
+        return 0;
+    }
+    virtual void stop()
+    {
+        DBGLOG ("TRACE: STOP (ActivityId: %d, %u)", activityId, rowRingBuffer->ordinality());
+        continueTracing=false;
+        waitSemaphore.signal();
+        join();
+    }
+    inline void enqueue(const void *row)
+    {
+        DBGLOG("TRACE enqueue: %p ActivityId: %d ordinality %u",row, activityId, rowRingBuffer->ordinality());
+        LinkThorRow(row);
+        rowRingBuffer->enqueue(row);
+    }
+};
+
+class CTracingThorDataLink :  implements CInterfaceOf<IThorDataLink>
+{
+private:
+    static const unsigned traceRowBufferSize=10;
+    static const unsigned dumpRowIntervalMS=600000;
+    IThorDataLink *thorDataLink;
+    IHThorArg *helper;
+    CRowTracerThread *rowTracerThread;
+
+public:
+    virtual void start()
+    {
+        DBGLOG("CTracingThorDataLink start %p", this);
+        thorDataLink->start();
+        rowTracerThread = new CRowTracerThread(helper,traceRowBufferSize,dumpRowIntervalMS,queryFromActivity()->queryActivityId());
```

Comment:
The downside of tracing on the same thread as you are gathering is that you won't get any tracing from a stalled job (which may be where you need it most). But I suspect from an efficiency point of view it may be the best option.